src/compiler/core.ts

```diff
@@ -409,6 +409,19 @@ namespace ts {
         return hasOwnProperty.call(map, key) ? map[key] : undefined;
     }
 
+    /** Gets a property in a Map, or if it doesn't exist, creates and returns it. */
+    export function getOrUpdate<T>(map: Map<T>, key: string, makeValue: () => T) {
+        return map[key] || (map[key] = makeValue());
+    }
+
+    /**
+     * Gets a value in an array, or if it doesn't exist, creates and returns it.
+     * It is the caller's responsibility to ensure that `key` is within the bounds of the array.
+     */
+    export function getOrUpdateArray<T>(arr: T[], key: number, makeValue: () => T) {

```

Comment:
It may make sense to do the same thing for this `makeValue` as the recommendation for `getOrUpdate`